tests/lambdas/leverDev.test.js

```diff
@@ -0,0 +1,79 @@
+const lambda = require('../../src/functions/lever')
+const got = require('got')
+const url = require('url')
+
+jest.mock('got')
+jest.mock('url')
+
+describe('different data, not production', () => {
+  let response
+
+  beforeAll(async () => {
+    url.URL.mockReturnValueOnce({
+      href: 'http://localhost:8000/meta.json',
+      origin: 'http://localhost:8000',
+      protocol: 'http:',
+      username: '',
+      password: '',
+      host: 'localhost:8000',
+      hostname: 'localhost',
+      port: '8000',
+      pathname: '/meta.json',
+      search: '',
+      hash: ''
+    })
+
+    url.URL.mockReturnValueOnce({
+      href: 'https://api.lever.co/v0/postings/yld?mode=json',
+      origin: 'https://api.lever.co',
+      protocol: 'https:',
+      username: '',
+      password: '',
+      host: 'api.lever.co',
+      hostname: 'api.lever.co',
+      port: '',
+      pathname: '/v0/postings/yld',
+      search: '?mode=json',
+      hash: ''
+    })
+
+    got.mockReturnValueOnce({
+      body: {
+        allJobIds: ['one', 'two', 'three']
+      }
+    })
+
+    got.mockReturnValueOnce({
+      body: [
+        {
+          id: 'one'
+        },
+        {
+          id: 'two'
+        },
+        {
+          id: 'three'
+        },
+        {
+          id: 'four'
+        },
+        {
+          id: 'five'
+        }
+      ]
+    })
+
+    response = await lambda.handler({
+      headers: {
+        authorization: 'basic foo_test_thingy'
+      }
```

Comment:
To remove the need to run these tests with hardcoded auth tokens and passing a header here we can use jest mocking to get around all of this by mocking the `Auth` module, I've done this [here](https://github.com/yldio/yld.io/compare/feat/lever-unit-tests?expand=1#diff-4772b57a44b209349e50cff83b53d1c8R1)

An  overview of what the auth and lambdas are doing.
```
const leverCallback = () => { // do the lambda stuff here }

const Auth = (event, callback) => {
    // do stuff with event data
   if (event data is ok) {
       return callback()  <---- remember this is is just the leverCallback in this instance
   }
}

const lambda = (event) => Auth(event, leverCallback)
```

With the line
```
jest.mock('../../src/functions/utils/auth', () => jest.fn((_, cb) => cb()))
```

All this says is, "ignore the first argument, just return the callback function and execute it", so we bypass the `Auth` module and just fire the lambdaCallback

Reply:
Yeah, I figured it's just going "eh, just give me the callback", but is there a technical reason why doing this is better? From my perspective it's doing the same thing, just a bit differently...

Reply:
Imagine we have 100 lambdas and we write tests for all of them and we pass the header you have here: 

```
headers: {
    authorization: 'basic foo_test_thingy'
}
```
Every single lambda test is testing our lambda functionality, great, but it's also testing the Auth function. So what's really happening is that we're writing 100 tests for lambdas and all of them test the Auth function. (This isn't want unit tests are about, thats for integration or e2e tests, something we've not touched on yet. It's really important you read up on the concepts of units tests.)

Anyway, imagine we change the internals of the Auth function to use a `Bearer` token or something changes which means the authorization header here is no longer working. What happens? All 100 tests fail and we have to change *all* of the lambda tests to reflect this change.

The difference between `passing the authorization header to the handler` and `mocking the Auth function` is that with the mock we are positive that the only  thing we're concerned about is the lambda under test, not the Auth function. It's really important to understand the difference here!

Reply:
And of course, we will be writing tests for the Auth function!